storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateEntry.java

```diff
@@ -0,0 +1,143 @@
+ * elements in the queue. When the queue is at capacity, we remove the first element to make space for the incoming batch.
+ */
+public class ProducerStateEntry {
+    public static final int NUM_BATCHES_TO_RETAIN = 5;
+    public final long producerId;
+    private final List<BatchMetadata> batchMetadata;
+    private short producerEpoch;
+    public int coordinatorEpoch;
+    public long lastTimestamp;
+    public OptionalLong currentTxnFirstOffset;
+
+    public ProducerStateEntry(long producerId) {
+        this(producerId, new ArrayList<>(), RecordBatch.NO_PRODUCER_EPOCH, -1, RecordBatch.NO_TIMESTAMP, OptionalLong.empty());
+    }
+
+    public ProducerStateEntry(long producerId, short producerEpoch, int coordinatorEpoch, long lastTimestamp, OptionalLong currentTxnFirstOffset) {
+        this(producerId, new ArrayList<>(), producerEpoch, coordinatorEpoch, lastTimestamp, currentTxnFirstOffset);
+    }
+
+    public ProducerStateEntry(long producerId, List<BatchMetadata> batchMetadata, short producerEpoch, int coordinatorEpoch, long lastTimestamp, OptionalLong currentTxnFirstOffset) {
+        this.producerId = producerId;
+        this.batchMetadata = batchMetadata;
+        this.producerEpoch = producerEpoch;
+        this.coordinatorEpoch = coordinatorEpoch;
+        this.lastTimestamp = lastTimestamp;
+        this.currentTxnFirstOffset = currentTxnFirstOffset;
+    }
+
+    public int firstSeq() {
+        return isEmpty() ? RecordBatch.NO_SEQUENCE : batchMetadata.get(0).firstSeq();
+    }
+
+
+    public long firstDataOffset() {
+        return isEmpty() ? -1L : batchMetadata.get(0).firstOffset();
+    }
+
+    public int lastSeq() {
+        return isEmpty() ? RecordBatch.NO_SEQUENCE : batchMetadata.get(batchMetadata.size() - 1).lastSeq;
+    }
+
+    public long lastDataOffset() {
+        return isEmpty() ? -1L : batchMetadata.get(batchMetadata.size() - 1).lastOffset;
+    }
+
+    public int lastOffsetDelta() {
+        return isEmpty() ? 0 : batchMetadata.get(batchMetadata.size() - 1).offsetDelta;
+    }
+
+    public boolean isEmpty() {
+        return batchMetadata.isEmpty();
+    }
+
+    public void addBatch(short producerEpoch, int lastSeq, long lastOffset, int offsetDelta, long timestamp) {
+        maybeUpdateProducerEpoch(producerEpoch);
+        addBatchMetadata(new BatchMetadata(lastSeq, lastOffset, offsetDelta, timestamp));
+        this.lastTimestamp = timestamp;
+    }
+
+    public boolean maybeUpdateProducerEpoch(short producerEpoch) {
+        if (this.producerEpoch != producerEpoch) {
+            batchMetadata.clear();
+            this.producerEpoch = producerEpoch;
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    private void addBatchMetadata(BatchMetadata batch) {
+        if (batchMetadata.size() == ProducerStateEntry.NUM_BATCHES_TO_RETAIN) batchMetadata.remove(0);
```

Comment:
`remove(0)` is expensive for `ArrayList`. This is likely the reason why `Queue` had been used previously. I suggest using `ArrayDeque` as the implementation and `java.util.Queue` or `java.util.Deque` as the interface.